test/test_modules.py

```diff
@@ -514,6 +514,109 @@ def _to_device(obj):
+            cudnn_greater_7_6 = torch.backends.cudnn.version() >= 7603
+            cudnn_greater_8 = torch.backends.cudnn.version() >= 8005
+
+        def _get_mem_formats(channels_last=False, channels_last_3d=False):
+            if channels_last:
+                return ([torch.contiguous_format, torch.channels_last],
+                        [torch.preserve_format, torch.contiguous_format, torch.channels_last])
+            elif channels_last_3d:
+                return ([torch.contiguous_format, torch.channels_last_3d],
+                        [torch.preserve_format, torch.contiguous_format, torch.channels_last_3d])
+            else:
+                return ([torch.contiguous_format],
+                        [torch.preserve_format, torch.contiguous_format])
+
+        # Check that at least one Tensor input has dim == n
+        def _check_dims(obj, n):
+            if isinstance(obj, torch.Tensor):
+                return obj.dim() == n
+            elif isinstance(obj, (tuple, list)):
+                return any(_check_dims(o, n) for o in obj)
+            else:
+                return False
+
+        # Called after _check_dims, when we know that >= 1 tensor can be converted to mem_format
+        def _to_mem_format(obj, mem_format):
+            if isinstance(obj, torch.Tensor):
+                d = obj.dim()
+                if ((mem_format == torch.channels_last and d != 4)
+                   or (mem_format == torch.channels_last_3d and d != 5)):
+                    return obj
+                return obj.to(memory_format=mem_format)
+            elif isinstance(obj, (tuple, list)):
+                return type(obj)(_to_mem_format(o, mem_format) for o in obj)
+            elif isinstance(obj, dict):
+                return {key: _to_mem_format(o, mem_format) for key, o in obj.items()}
+            else:
+                return obj
+
+        def _check_out_mem_format(output, input_mem_format, module_mem_format):
+            if isinstance(output, torch.Tensor):
+                d = output.dim()
+                if (d == 4 and ((input_mem_format == torch.channels_last and (not is_conv or cudnn_greater_7_6))
+                                or (module_mem_format == torch.channels_last and is_conv and cudnn_greater_7_6))):
+                    self.assertTrue(output.is_contiguous(memory_format=torch.channels_last))
+                elif (d == 5 and ((input_mem_format == torch.channels_last_3d and (not is_conv or cudnn_greater_8))
+                                  or (module_mem_format == torch.channels_last_3d and is_conv and cudnn_greater_8))):
+                    self.assertTrue(output.is_contiguous(memory_format=torch.channels_last_3d))
+                else:
+                    self.assertTrue(output.is_contiguous())
+            elif isinstance(output, (tuple, list)):
+                [_check_out_mem_format(o, input_mem_format, module_mem_format) for o in output]
+            else:
+                return
```

Comment:
For these helper functions, it looks like there is a similar pattern of "traverse a data structure containing tensors and do something" - is it possible to refactor these to use `_traverse_obj()` instead?